### ISSUE TITLE
__dirname intellisense adding extra underscore

### ISSUE THREAD

**Issue by vsccarl** [Microsoft/vscode#4680]
- VSCode Version: 0.10.12-alpha
- OS Version: Windows 10 and Ubuntu 15.10

Steps to Reproduce:

1. Open a JavaScript file
2. Type underscore _ to bring up intellisense

Actual:
Triple underscore: ___dirname
![tripleunderscoredirname](https://cloud.githubusercontent.com/assets/12900364/14054265/1f7d2bbe-f295-11e5-9d54-3cc5f4b47078.JPG)

Expected:
Only 2 underscores: __dirname

**Comment by jrieken** [comment]
I'd reckon that the word def has an issue

**Comment by dbaeumer** [comment]
The proposed ___dirname is a word in that file (observe the file icon in front). The JS intellisense proposes words in the file as well.

If you install the node.d.ts file using tsd install node and then do _<code complete> you will see 

![capture](https://cloud.githubusercontent.com/assets/1931590/14669619/e779779a-06e9-11e6-9f4e-0f6d53205ffe.png)

because the JS language service then knows about __dirname

**Comment by cllivingston** [comment]
The extra underscore appears to happen with other symbols of the `__name` format in the global namespace.

![vscode-bug](https://cloud.githubusercontent.com/assets/3939610/14968939/770c362a-1074-11e6-98b4-6b34eba83dad.png)

This is in 1.0.0

**Comment by vilic** [comment]
@dbaeumer Steps to reproduce in 1.0:

1. Open a JavaScript file, and type an expression like `__foo;`.
2. Now type `_` and the intelisense suggests `___foo`.

**Comment by dbaeumer** [comment]
@vilic now I see. 
```js
var __foo;
_<code assist>
```
works but
```js
__foo;
_<code assist>
```
doesn't.

@jrieken: The proposal is a textual proposal an proposes ___foo; (three underlines). Any hint what to tweak here?

**Comment by dbaeumer** [comment]
![capture](https://cloud.githubusercontent.com/assets/1931590/15468925/4abd37da-20e7-11e6-9056-73d7f6ea0b9c.png)

**Comment by jrieken** [comment]
@dbaeumer This is usually an issue with the word definition - textual completion goes with the word def of the current mode and uses that to determine what portion of a word to overwrite/replace with a suggestion.

**Comment by dbaeumer** [comment]
@jrieken: the regular expression is
```js
/(-?\d*\.\d\w*)|([^\`\~\!\@\#\%\^\&\*\(\)\-\=\+\[\{\]\}\\\|\;\:\'\"\,\.\<\>\/\?\s]+)/g
```
which actually fully matches __foo according to https://regex101.com/r/eP0nI9/1
Any other idea?

**Comment by jrieken** [comment]
catch22 - that's actually not a textual completion but the TypeScript Salsa feature that proposes all symbols on top of semantic suggestions - note differences on the icon. The `___foo` suggestion is delivered as part of the other suggestions. 

<img width="644" alt="screen shot 2016-05-23 at 15 33 50" src="https://cloud.githubusercontent.com/assets/1794099/15472525/dabf7050-20fb-11e6-8e5d-c8a359bf5b6d.png">

**Comment by dbaeumer** [comment]
@jrieken thanks. I opened https://github.com/Microsoft/TypeScript/issues/8778 to track this in TS land.

**Comment by kumarharsh** [comment]
@dbaeumer it seems like the issue has been fixed in TS, but has not come into vscode (latest insider) yet. I'm still getting one extra underscore on my variable names.

**Comment by dbaeumer** [comment]
@kumarharsh that is correct. The fix is in TS 2.0 which is not official yet. However you can switch to 2.0 RC version using the tsdk setting.
